### PR TITLE
[FLINK-15467][task] Wait for sourceTaskThread to finish before exiting from StreamTask.invoke [1.10]

### PR DESCRIPTION
**Pull request by AHeise**
(Unchanged backport of https://github.com/apache/flink/pull/13000)

## What is the purpose of the change

Wait for the LegacySourceFunctionThread to finish when cancelling the task.
This prevents premature freeing of resources, particularly LibraryCache (see [FLINK-15467](https://issues.apache.org/jira/browse/FLINK-15467) discussion).

Implemented by waiting in `SourceStreamTask.invoke()` for the `sourceThread` to complete.

There were several existing tests failures (listed below) which were fixed by completing the future if thread isn't alive; or by returning a completed future if a task is failing.
## Verifying this change

* Added unit test: `SourceStreamTaskTest.testWaitsForSourceThreadOnCancel`

* Tested manually on a local cluster

* Existing tests covering different scenarios of task exit and cancellation: `InterruptSensitiveRestoreTest`, `StreamTaskTest.testEarlyCanceling`, `FastFailuresITCase`

## Does this pull request potentially affect one of the following parts:

* Dependencies (does it add or upgrade a dependency): no

* The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no

* The serializers: no

* The runtime per-record code paths (performance sensitive): no

* Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no

* The S3 file system connector: no

## Documentation

* Does this pull request introduce a new feature? no

* If yes, how is the feature documented? no

### CODE REVIEW

**Review by pnowojski — APPROVED** [review]
I presume this is 1to1 backport without any changes? LGTM if travis is green. One thing to note, I don't know how are we standing with end to end tests coverage. For older releases (and travis) we might not have them executed (lack of nightly builds) and it would be good to verify e2e tests as well.

**Comment by AHeise** [comment]
@flinkbot run travis

**Comment by pnowojski** [comment]
@flinkbot run travis

**Comment by AHeise** [comment]
@flinkbot run travis

**Comment by AHeise** [comment]
@flinkbot run travis

**Comment by AHeise** [comment]
Travis on Flink CI doesn't like to pick up the right commit.

On my private travis, there were a few timeouts, but it looks green to me ;) https://travis-ci.org/github/AHeise/flink/builds/727356614

**Pull request merged**
